feat(posts): six draft posts from GopherCon UK 2026 talk (#311)
kakkoyun/me: content/posts/why-go-cant-be-monkey-patched.md

---
title: "Why Go can't be monkey-patched (and what people do about it)"
description: "Go is structurally hostile to zero-touch instrumentation, from the missing classloader to a goroutine-local storage hack that patches the runtime at compile time."
date: 2026-09-01T00:00:00Z

> publishDate: 2026-09-01T00:00:00Z

Author: Sep 1 is a **Tuesday, and it already has two posts on it** — "How I Use Agents, Part 1" (#209) and benchmarks 1/5 (#312). Three posts, one day.

**Post ① → Fri Oct 2:**

```suggestion
date: 2026-10-02T00:00:00Z
publishDate: 2026-10-02T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: "Why Go can't be monkey-patched (and what people do about it)"
description: "Go is structurally hostile to zero-touch instrumentation, from the missing classloader to a goroutine-local storage hack that patches the runtime at compile time."
date: 2026-07-28T00:00:00Z

> publishDate: 2026-07-28T00:00:00Z

Author: **Jul 28 is eight days in the past.** The `publishDate` added here is a copy of the old `date`, so the frontmatter is now well-formed and still unpublishable — the moment `draft: true` comes off, this goes live backdated to July, which is the exact outcome the suggestion was meant to prevent. Same in all six files (Jul 28 – Aug 3).

**Post ① → Fri Oct 2:**

```suggestion
date: 2026-10-02T00:00:00Z
publishDate: 2026-10-02T00:00:00Z
```

Everything else this commit did to the frontmatter is right: `description` and `categories` filled in on the three files that lacked them, tags in list form with `blog` added, `showToc`/`tocOpen` replacing `toc`, redundant `slug` gone, full RFC 3339 timestamps.

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: "Why Go can't be monkey-patched (and what people do about it)"
date: 2026-07-28

> draft: true

Author: **Post ① of the drop → Friday, September 25.**

None of the six posts here has a `publishDate` at all, only `date`. Without one, Hugo treats `date` as the publish date, so the moment `draft: true` comes off, all six go live at once on their next build — dated late July, in the past. That's the failure mode this suggestion prevents:

```suggestion
date: 2026-09-25T00:00:00Z
publishDate: 2026-09-25T00:00:00Z
draft: true
```

Note the timestamp format too. `date: 2026-07-28` with no time parses as midnight UTC, which happens to work, but every other post in the repo uses the full RFC 3339 form and the publish cron's date comparison is easier to reason about when they match.

---
_Generated by [Claude Code](https://claude.ai/code)_

tags: ["go", "observability", "opentelemetry", "ebpf", "compile-time-instrumentation"]
categories: ["engineering"]
slug: "why-go-cant-be-monkey-patched"
toc: true
---

If you've spent time on Java or Python instrumentation, the zero-touch story is obvious: intercept the bytecode at load time, or swap out a method at runtime, or use an agent that the JVM already knows how to host. Go has none of that. This isn't a gap waiting to be filled — it's a consequence of how Go was designed.

### The structural problem

Go compiles to native machine code. When you run `go build`, you get a self-contained binary. There is no bytecode layer, no classloader, no intermediate representation sitting between your code and the CPU.

That matters because it closes off the standard instrumentation escape hatches:

- No bytecode means you can't rewrite classes at load time. Java agents work because the JVM can intercept `ClassLoader.defineClass` and transform bytecodes before they execute. Go has no equivalent — the compiler runs on your machine, not on the server.
- No classloader means there's no intercept point between "file on disk" and "code executing."

> - Go links statically by default. The `LD_PRELOAD` trick — the classic Linux mechanism for injecting a shared library into any process — requires the dynamic linker to run, and Go's internal linker doesn't invoke it. To use `LD_PRELOAD` with a Go binary, you have to force external linking with `-linkmode=external`. That's not a supported feature; it's a workaround with its own sharp edges. Dynatrace's OneAgent documentation says outright: static Go binaries with cgo are unsupported.

Author: "Static Go binaries **with cgo** are unsupported" is self-contradictory — static Go binaries are typically the `CGO_ENABLED=0` ones; cgo is what usually forces *dynamic* linking (and gives Dynatrace its libc injection point). Dynatrace's actual limitation is the inverse: statically linked binaries (no dynamic loader, e.g. `CGO_ENABLED=0` or musl-static) can't be injected into. Reword to "statically linked Go binaries — the `CGO_ENABLED=0` default for containers — are unsupported" and re-check their docs page for the current phrasing while you're there.

---
_Generated by [Claude Code](https://claude.ai/code)_

- There is no general runtime hook API. Go has an internal `exithook` mechanism, but it's scoped strictly to program termination — it's not a general interception surface.

What about `go:linkname`? It lets packages access unexported runtime symbols, and you'll find it used in tracers. But the Go runtime source contains this comment in `proc.go`:

> "gopark should be an internal detail, but widely used packages access it using linkname. Notable members of the hall of shame include…"

> The Go team added those stubs reluctantly, because enough widely-used packages had already started abusing them. That's not a hook point — it's a "we can't remove this without breaking half the ecosystem" accommodation.

Author: `ecosystem` is on the Vale banned list (error level) — this is the one hard vocab hit in the whole eleven-post drop.

```suggestion
The Go team added those stubs reluctantly, because enough widely-used packages had already started abusing them. That's not a hook point — it's a "we can't remove this without breaking half the module graph" accommodation.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

The way DataDog's compile-time tool, Orchestrion, solves this is striking. It injects a synthetic field directly into the Go runtime's internal `g` struct — the struct that represents each goroutine:

```yaml
# internal/orchestrion/gls.orchestrion.yml
join-point:
  struct-definition: runtime.g
advice:
  - add-struct-field:
      name: __dd_gls_v2
      type: any
```

This YAML aspect tells the compiler to add `__dd_gls_v2 any` to `runtime.g`. Every goroutine gets a copy. Then Orchestrion injects `go:linkname` functions to get and set that field:

```go
//go:linkname __dd_orchestrion_gls_get __dd_orchestrion_gls_get.V2
var __dd_orchestrion_gls_get = func() any {
    return getg().m.curg.__dd_gls_v2
}

//go:linkname __dd_orchestrion_gls_set __dd_orchestrion_gls_set.V2
var __dd_orchestrion_gls_set = func(val any) {
    getg().m.curg.__dd_gls_v2 = val
}
```

And to prevent memory leaks, a cleanup hook into `goexit1`:

```go
// Injected into runtime.goexit1:
getg().__dd_gls_v2 = nil
```

The consumer side in dd-trace-go checks whether Orchestrion is present at init time and falls back to a no-op if it isn't.

> This is genuinely impressive engineering. It's also a sign of how desperate things are: you're patching the Go runtime struct layout at compile time, using `go:linkname` for variable aliasing (which broke between Go 1.22 and 1.23, tracked in golang/go#72032), and relying on a YAML file that references the Go team's own `HACKING.md`. The Go team considers this access pattern deeply unofficial.

Author: Double-check `golang/go#72032` — the well-known linkname lockdown (the Go 1.23 restriction that broke pull-style `go:linkname` uses) is tracked under **golang/go#67401** ("cmd/link: lock down future uses of linkname"). If #72032 is a narrower issue specific to variable-alias linkname or Orchestrion's case, keep it but consider citing #67401 alongside as the umbrella — readers who know this saga know it by that number. I can't reach GitHub from this environment to confirm what #72032 actually is.

---
_Generated by [Claude Code](https://claude.ai/code)_
